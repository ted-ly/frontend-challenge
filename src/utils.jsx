import { useState, useCallback } from 'react';

export const initialState = {
    name: '',
    password: '',
    color: '',
    email: '',
    terms: false,
};

export const onClickNext = (to, setPathname) => {
    setPathname(to);
    window.history.pushState(null, '', to);
};

export const onClickBack = (from, setPathname) => {
    setPathname(from);
    window.history.pushState(null, '', from);
}; 

export const onSubmit = async (formData, setPathname) => {
    const resp = await fetch('http://localhost:3001/api/submit', { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(formData) });
    if (resp.status >= 400) {
        window.history.pushState(null, '', '/error');
        setPathname('/error');
    } else {
        window.history.pushState(null, '', '/success');
        setPathname('/success');
    }
};

export const createOnClickRestart = ({ setFormData, setPathname }) => useCallback(() => {
    window.history.pushState(null, '', '/');
    setPathname('/');
    setFormData({...initialState});
}, []);

export const createOnFormDataChange = ({ setFormData, formData }) => useCallback((e) => {
    const name = e.target.name;
    const value = e.target.value;
    const newFormData = {
        ...formData,
        [name]: name ==='terms' ? !formData.terms : value,
    };

    setFormData({...newFormData});
}, [formData]); 

export const useForm = () => {
    const [formData, setFormData] = useState(initialState);
    const [pathname, setPathname] = useState('/');
    const onFormDataChange = createOnFormDataChange({ formData, setFormData });
    const onClickRestart = createOnClickRestart({ setFormData, setPathname });

    return { onClickRestart, onFormDataChange, formData, setPathname };
};

export const toCapitalized = (str) => {
    return str.slice(0, 1).toUpperCase() + str.slice(1);
}

export const toTransformText = (text, key) => {
    if (key === 'terms') {
        return text ? 'Agreed' : 'Disagreed';
    }

    if (key === 'password') {
        return '*'.repeat(text.length);
    }

    return text;
}; 