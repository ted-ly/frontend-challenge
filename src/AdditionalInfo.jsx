import React, { useCallback, useState, useEffect } from 'react';
import Button from './Button';
import Input from './Input';
import Step from './Step';
import Select from './Select';
import { onClickBack, onClickNext } from './utils';

export default function AdditionalInfo({ formData, onFormDataChange, setPathname }) {
    const [options, setOptions] = useState([]);
    const [isLoading, setIsLoading] = useState(false);

    useEffect(() => {
        async function fetchData() {
            if (options.length < 1) {
                setIsLoading(true);
                const options = await fetch('http://localhost:3001/api/colors');
                const jsonOptions = await options.json();
                setOptions([...jsonOptions]);
                setIsLoading(false);
            }
        }
        fetchData();
    }, [options]);

    const next = useCallback(() => {
        onClickNext('/confirmation', setPathname);
    }, []);
    const back = useCallback(() => {
        onClickBack('/', setPathname);
    }, []);
    return (
        <Step title='Additional Info' to='more-info' isLoading={isLoading}>
            <Select name='color' options={options} onChange={onFormDataChange} />
            <div className='terms-container'>   
                <Input type='checkbox' name='terms' checked={formData.terms} onChange={onFormDataChange} />
                <label className='terms-text'>I agree to <a href=''>Terms And Conditions</a></label>
             </div>
            <div className='button-container'>
                <Button onClick={back}>Back</Button>
                <Button onClick={next} className='next-btn'>Next</Button>
            </div>
        </Step>
    )
}