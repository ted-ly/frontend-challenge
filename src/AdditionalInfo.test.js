import renderer from 'react-test-renderer';
import AdditionalInfo from './AdditionalInfo';
import { initialState } from './utils';

describe('AdditionalInfo', () => {
    it('should match with snapshot', () => {
        const component = renderer.create(<AdditionalInfo formData={initialState} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});