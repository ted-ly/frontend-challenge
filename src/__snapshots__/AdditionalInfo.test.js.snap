// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInfo should match with snapshot 1`] = `
<div
  className="step-container"
>
  <div
    className="step-title"
  >
    Additional Info
  </div>
  <select
    name="color"
  >
    <option
      value=""
    >
      Select Your Favorite Color
    </option>
  </select>
  <div
    className="terms-container"
  >
    <input
      checked={false}
      name="terms"
      type="checkbox"
    />
    <label
      className="terms-text"
    >
      I agree to 
      <a
        href=""
      >
        Terms And Conditions
      </a>
    </label>
  </div>
  <div
    className="button-container"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Back
    </button>
    <button
      className="next-btn"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</div>
`;
